lib/compatibility.bzl

```diff
@@ -1,64 +1,96 @@
+    name = targets_name.replace("//", "").replace(":", "_").replace("/", "_").replace(".", "_")
+    return name
 
-# Another alternative:
-# def compatibility(any_of = [], none_of = [], all_of = None):
+def _maybe_make_unique_incompatible_value(name):
+    """Creates a `native.constraint_value` which is "incompatible."
 
+    When composing selects which could all resolve to "incompatible" we need distinct labels.
+    This will create a constraint_value with the given name, if it does not already exist.
+
+    Args:
+      name: A target name to check and use.
+    """
+    if not native.existing_rule(name):
+        native.constraint_value(name = name, constraint_setting = INCOMPATIBLE_SETTING)
 
 def _none_of(settings):
-    return selects.with_or({"//conditions:default": [], tuple(settings): [INCOMPATIBLE_TARGET]})
+    """Create a `select()` which matches none of the given config_settings.
+
+    Any of the settings will resolve to an incompatible constraint_value for the
+    purpose of target skipping.
+
+    Args:
+      settings: A list of `config_settings`.
+
+    Returns:
+      A native `select()` which maps any of the settings to the incompatible target.
+    """
+    compat_name = " incompatible with " + _get_name_from_target_list(settings)
+    _maybe_make_unique_incompatible_value(compat_name)
+    return selects.with_or({"//conditions:default": [], tuple(settings): [":" + compat_name]})
 
 def _any_of(settings):
-    no_match_error = "Didn't match any conditions." # TODO(trybka): concat setting names?
-    return selects.with_or({tuple(settings): [],}, no_match_error = no_match_error)
+    """Create a `select()` which matches any of the given config_settings.
+
+    Any of the settings will resolve to an empty list, while the default condition will map to
+    an incompatible constraint_value for the purpose of target skipping.
+
+    Args:
+      settings: A list of `config_settings`.
+
+    Returns:
+      A native `select()` which maps any of the settings an empty list.
+    """
+    compat_name = " compatible with any of " + _get_name_from_target_list(settings)
+    _maybe_make_unique_incompatible_value(compat_name)
+    return selects.with_or({tuple(settings): [], "//conditions:default": [":" + compat_name]})
 
 def _all_of(settings):
-    settings_name = "_and_".join([s.split(":")[1] for s in settings])
-    name = settings_name.replace(":", "-").replace("//", "").replace("/", "_")
-    if not native.existing_rule(name):
-        selects.config_setting_group(name = name, match_all = settings)
-    no_match_error = "Didn't match all conditions." # TODO(trybka): concat setting names?
-    return select({":" + name: []}, no_match_error=no_match_error)
+    """Create a `select()` which matches all of the given config_settings.
+
+    All of the settings must be true to get an empty list. Failure to match will result
+    in an incompatible constraint_value for the purpose of target skipping.
+
```

Comment:
Could you expand this to talk about `constraint_value` targets not working with `all_of` because of https://github.com/bazelbuild/bazel/issues/13047?

As an work around, folks should use something like this if they need to mix `constraint_value` and `config_setting` targets:
```python
target_compatible_with = [":constraint1", "constraint2"] + compatbility.all_of([":config1", ":config2"]),
```

Reply:
I commented on the bug but I'm really hoping that bug is fixable without a major effort. It's ultimately a silly circular dependency problem. Those are always awkward, but conceptually not *too* hard to think about ways of short circuiting.